Add 'entity_picture' property to DarkSkySensor component

aronsky opened the pull request:
**Description:**


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

**Example entry for `configuration.yaml` (if applicable):**
```yaml

```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

aronsky committed:
Add 'entity_picture' property to DarkSkySensor component

homeassistant commented:
Hi @aronsky, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5948).

Once you do that we will be able to review and accept this pull request.

Thanks!

aronsky commented:
I attempted to run `tox` tests, and had one failure (in the `test_darksky.py` module, suspicious!) - but this failure seems to happen prior to my changes, as well. Looks like something about a mock `load_fixture`, which has nothing to do with the changes in this pull request.

pvizeli commented:
I don't think that is a good idea to add base64 icons to code. But I can also not say what the alternative sultion is. I'm not a frontend developer. CC @balloob

aronsky committed:
Restructured the long lines to fit PEP8 guidelines

aronsky commented:
@pvizeli I agree that it's not perfect, far from it. I just wanted it to be both self-contained, and not dependent on 3-rd party repositories, which might change and/or disappear at any moment, breaking the module.

I looked around on the Dark Sky website, hoping to find some sort of permalinks to the icons, but their official icons are dynamic and use Javascript - I don't know how to cleanly integrate those into Home Assistant.

If there's a more elegant solution, I'll be happy to do it.

fabaff commented:
One idea of the `weather` component was to be able to display the conditions in a visual way. Unfortunately the frontend part is not done yet.

rpitera commented:
@aronsky Could you use the Skycons library?  It's self contained JS, public domain and created for DarkSky.  Plus the icons are animated.

https://github.com/darkskyapp/skycons

aronsky commented:
@rpitera - yep, I saw those icons. They definitely look cool - but integrating them requires changes in the front-end, and without that - I have no idea how a specific module can influence the front-end extensively (and this is required, since the front-end doesn't currently support Javascript-based icons)

aronsky committed:
Merge branch 'dev' of github.com:home-assistant/home-assistant into dev

balloob commented on homeassistant/components/sensor/darksky.py:
```diff
@@ -97,6 +97,346 @@
 }
 
+CONDITION_PICTURES = {
+    'clear-day': ('data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZ'
```
Adding them here will mean they end up in the state machine, in the database and slow down initial load of the app. Not allowed.

Please remove all the inline images and instead add them to a new folder in the static folder here: https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/frontend/www_static/images .

aronsky replied:
Placing specific weather icons in the images folder seems a little out of place. I looked more at the code, and I think I found the relevant changes that need to be made to the frontend component in order to support weather icons. I'm gonna do that, and once it's done - every weather sensor will be able to display standardizes Material Design icons.

balloob commented:
I'm btw a big fan of the skycons. We should consider adding those to the frontend.

aronsky committed:
Switched to statically stored PNG weather icons

aronsky committed:
Added missing  URL prefix

aronsky committed:
Added SVG weather icons. Switched to a different-style rain icon.

aronsky commented:
OK. So, updating the frontend proved to be more complicated than I thought. I have, in fact, made a small patch to the frontend, that would show the state's `icon` property (instead of the currently used `entity_picture`). But this behavior is only desirable for weather forecast sensors. Unfortunately, by the time the state arrives to the frontend, there's nothing to differentiate it from other sensors, even other monitored conditions from the weather sensor (such as the temperature, which should display the value, not the icon). As a temporary hack, I enabled this only for sensors that have an icon containing the word weather, but obviously it's not a proper solution.

I ended up doing what @balloob suggested for now, and will try to come up with a more robust solution to the frontend in the future.

Speaking of the frontend, what are the guidelines to making changes to that component, since it compiles into files that are part of this repo? Do I make a pull request in the frontend's repository, and once it is accepted, compile and make a pull request here?

balloob commented on homeassistant/components/sensor/darksky.py:
```diff
@@ -262,6 +291,9 @@ def get_state(self, data):
         if state is None:
             return state
 
+        if "summary" in self.type:
```
Do you mean `self.type == "summary"` ?

aronsky replied:
Nope, since in DarkSky the sensor supports several types that are forecasts: `summary`, `minutely_summary`, `hourly_summary`, and `daily_summary`. I think `"string" in variable` is more readable/Pythonic than `variable.find("string") >= 0`.

balloob commented on homeassistant/components/sensor/darksky.py:
```diff
@@ -188,6 +203,17 @@ def unit_system(self):
+    @property
+    def entity_picture(self):
+        """Return the entity picture to use in the frontend, if any."""
+        if self._icon is None or "summary" not in self.type:
```
You set icon to an empty string . So instead you should check just `if self._icon`

aronsky replied:
You mean `if not self._icon` :)
Fixing.

aronsky replied:
Rechecked. I set `self._icon` to `None` in the constructor. I only set it to a value (or empty string, if `getattr` fails) for specific types of the sensor (namely, all types that have `summary` in their name). So the first part of the condition is correct - I'm looking at cases when `self._icon` hasn't been set since construction time. You could say that the second part of the condition is redundant - if you think it doesn't belong there, I'll drop it.

balloob commented:
You just make PRs against the frontend repo. I add new builds from frontend to this repo.

Please add your images in a subfolder under images.

Also a heads up that this is the last change that we will allow for the darksky sensor. Future efforts will have to be focussed on the weather component. This sensor will be removed in a future release.

aronsky commented:
I had no idea there's a weather component in the works - I didn't see anything in the documentation. I'll look into it (it does make sense to have those unified somehow, as opposed to different sensors with different images and weather states).

I'll move the images to a subfolder.

And finally, regarding the frontend repo. I understand that this might be a headache to keep track of, but any chance you can tag frontend commits with versions matching the home-assistant tags that use those commits? It's a lot less clear on which version to work on in that repo.

aronsky committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into dev

aronsky committed:
Moved darksky images to an appropriate subfolder. Changed double to single quotes to match style.

aronsky commented on homeassistant/components/sensor/darksky.py:
```diff
@@ -188,6 +203,17 @@ def unit_system(self):
+    @property
+    def entity_picture(self):
+        """Return the entity picture to use in the frontend, if any."""
+        if not self._icon is None or 'summary' not in self.type:
```
Oops

aronsky committed:
Oops, this  shouldn't have slipped in here

nodinosaur committed:
Added a config flag[Boolean] to declare if SenseHAT is attached  (#5883)

* Added a config flag[Boolean] to declare if SenseHAT is attached to RaspberryPi

* Hound found a line too long violation - fix

aequitas committed:
Reconnect robustness, expose connection state. (#5869)

* Reconnect robustness, expose connection state.

- Expose connection status as rflink.connection_status state.
- Handle alternative timeout scenario.
- Explicitly set a timeout for connections.
- Error when trying to send commands if disconnected.
- Do not block component setup on gateway connection.

* Don't use coroutine where none is needed.

* Test disconnected behaviour.

* Use proper conventions for task creation.

* Possibly fix test race condition?

* Update hass import style

balloob commented:
Every Home Assistant commit already contains a version of Home Assistant Polymer via a submodule: https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/frontend/www_static

That being said, you still always have to work on the latest version no matter what commit it is based off.

balloob commented:
Can you remove the png versions since they are not being used ?

balloob commented:
You also broke the tests.

postlund committed:
Add fake support for turn on/off for Apple TV (#5962)

* Add fake support for turn on/off for Apple TV

When the device is "turned off", no requests are sent to the device.
When the setting "start_off" is set to true, the device starts in off
state.

* Fix async comments

* Clean up supported features

aronsky committed:
Removed unnecessary PNG icons for DarkSky

homeassistant commented:
Hello @aronsky,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`01f346ccee1f62f354b0ee239ee9c42c38a43437`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;5948&#x2F;commits&#x2F;01f346ccee1f62f354b0ee239ee9c42c38a43437) - No email found attached to the commit.
- [`b2cbde31945d9bf0e9ccc87c3e9259da781cde1a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;5948&#x2F;commits&#x2F;b2cbde31945d9bf0e9ccc87c3e9259da781cde1a) - No email found attached to the commit.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @aronsky,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`01f346ccee1f62f354b0ee239ee9c42c38a43437`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;5948&#x2F;commits&#x2F;01f346ccee1f62f354b0ee239ee9c42c38a43437) - No email found attached to the commit.
- [`b2cbde31945d9bf0e9ccc87c3e9259da781cde1a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;5948&#x2F;commits&#x2F;b2cbde31945d9bf0e9ccc87c3e9259da781cde1a) - No email found attached to the commit.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

aronsky commented:
I'm not sure what broke the build... Any chance it has been broken in the `dev` branch that I merged into my code? It seems like it, because that's the first commit that seems to break the build. I merged the current `dev` branch, and hopefully it'll work - I'm not on my work computer, and can't run `tox` from here. If it still fails, it'll have to wait until tomorrow.

P.S. Sorry for the mess with the commits without e-mail addresses, I haven't used my home computer for development before, and didn't notice that git hasn't had an e-mail set up.

rpitera commented:
Seeing that this is using a subfolder of SVG icons, would this support putting in your own SVG icons as long as the naming matched the DarkSky naming conventions?   Also, if you replaced them, would they get overwritten on subsequent HA updates?   I'm assuming no, since this will be the final update (barring hot-fixes of course) allowed for this platform.

aronsky commented:
@rpitera yes, if this pull request is accepted, you'll be able to switch to your own SVGs, as long as you replace the files (i.e., the names match). Your assumption about the subsequent updates sounds correct to me, but your guess is as good as mine in that regard.

rpitera commented:
Thanks!

balloob commented:
It seems that you have added some commits to your branch unrelated to this change. Please clean it up.

aronsky commented:
Hi @balloob, I'm not sure which commits you're referring to. There were a couple of commits I've pushed without properly signing them with my email, and I had to do some git history rewriting to fix that. Other than that, there are only commits relevant to the change, and the merges from dev I did from time to time to stay up to date with the code. If that's not something I should've done, I'll avoid it in the future, but at this point I don't know how to unravel those merges - and not sure how critical it is, since those are all commits accepted into the `dev` branch, anyway.

Do you prefer that I redo my changes in a separate, clean pull request, and abandon this one?

balloob commented:
Have a look at commits or files changed to see what I mean:

![screen shot 2017-02-19 at 15 44 17](https://cloud.githubusercontent.com/assets/1444314/23107953/4e11c104-f6ba-11e6-8f69-f40e8e4f8d62.png)

aronsky commented:
Yup, that's the mess with the authoring I referred to previously, sorry about that. I'm not sure how to fix that properly - if I start afresh, rebase my `dev` branch on the original commit where it diverged from the upstream repo, cherry-pick the commits with my changes, and then push - will it clean everything up and avoid the need for a new pull request?

Again, sorry about the mess.

balloob commented:
Yes, that will work. You will have to force-push to overwrite the bad commits in your current branch.

To be safe, you can also create a new branch that branches off latest dev and cherry-pick your changes.

aronsky commented:
That sounds like the best idea. Will there be an option to update this pull request to point to the new branch, or do I check its correctness and then force-push it onto the current branch in question?

balloob commented:
Can't change the source branch. It will be easiest to just close this PR and open a new one.

aronsky commented:
I'll do that. Once again, sorry about the mess with the branches and stuff. The new pull request, #6141, is just one clean commit :)
